_posts/2021-07-13-JEV.md

```diff
@@ -0,0 +1,822 @@
+
+So by dividing `128 / 32 = 4`, meaning we can do operations on 4 ints at a time.
+
+```java
+final vector.VectorSpecies<Integer> species=vector.IntVector.SPECIES_128;
+```
+
+Sweet, but do I need to do all these tedious calculations everytime I want to do me some vectors?
+
+Luckily, no. The API contains a couple of convenience constants that do this for you.
+
+```java
+final vector.VectorSpecies<Integer> species = vector.IntVector.SPECIES_PREFERRED;
+```
+
+###### Read input into SIMD register(s)
+
+So next, we have to transform our plain old `int[]` into the Vector version. Or, in other words, move the data into the
+special SIMD registers.
+
+The funny thing about this part though, is that it seems very counterintuitive from the performance point of view. I
+mean, we are literally copying the data into smaller chunks, instead of iterating over an array?? At the very least you
+use double the memory size. 
+
+The problem is, our naïveté assumes that an array stays where it is, but know this, SIMD or not, a CPU only deals with
+registers.
+
+```java
+final int offset = 0;
+IntVector vector = IntVector.fromArray(species, rgb, offset);
+```
+
+Here we create a _Vector_ of the type and width we determined with the `species` before.
+
+Meaning, if `species == 128 bytes wide`, and we are dealing with `integers`, then `vector` will contain the
+first `4 ints` from the `rgb` array, at the specified `offset`.
+
+###### Do operation(s) on SIMD register(s)
+
+Now that the data is where we want it to be, we can unleash our CPU prowess upon it!!!
+
+As you will no doubt discover when you press ctrl + space on `vector.*`, there's a lot of fun things you can do with the
+vector. These of course are but a sane subset of SIMD, and instead of whining about it, be grateful (and send John Rose
+& Paul Sandoz a gift basket).   
+
+```java
+IntVector result = vector.add(1);
+```
+
+This step is self-explanatory.
+
+###### Write SIMD register(s) back to input
+
+```java
+result.intoArray(rgb, offset);
+```
+
+And finally, we write the updated values back to the array. The way we do it here isn't really the 'proper' way, but it
+works nonetheless. By proper, I mean writing the values back to the input array. It's much better to write to an
+intermediate array (as we might explain later). 
+
+###### Hello hello
+
+This is what our end result looks like:                                                      
+
+```java
+final VectorSpecies<Integer> species = IntVector.SPECIES_PREFERRED;
+int offset = 0;
+for (offset = 0; offset < rgb.length; offset += species.length()) {
+    final IntVector vector = IntVector.fromArray(species, rgb, offset);
+    final IntVector result = vector.add(1);
+    result.intoArray(rgb, offset);
+}
+```
+
+A casual observer might...observe, that we are still looping over our array. This is a necessary evil. Computers are
+magical, but not THAT magical. Maybe someday there will be a single instruction that operates on any length of data.
```

Comment:
There is, it is called `Arrays.parallelSetAll(array, element -> element + 1)`. Optimizes to 1 instruction, guaranteed!